templated-this.md

```diff
@@ -1,151 +1,315 @@
+| `void f(T const volatile& this)`  | `void f() const volatile&`  |            | 
+| `void f(T const volatile&& this)` | `void f() const volatile&&` |            | 
+
+*Notes:*
+
+- *[value]*: whether passing by value should be allowed is debatable, but seems
+  obviously desired for completeness and parity with inline friend functions.
+- The interpretation of `this` in the method body differs, but only one
+  definition for a given signature may be present, eg. one may define at most
+  one of `void f()&`, or `void f(T& this)` or `void f()`, the first and last
+  already being exclusive of one another.
+
+
+### How does templated `this` work?
+
+Using existing deduction rules for template parameters, which will deduce the
+type of `this` to something in the above table.
+
+
+### What does `this` mean in the body of a method?
+
+It behaves exactly as a regular parameter declared in the same way.
+
+
+### Constructors
+
+No exceptions to above rules. If a particular constructor signature is not
+allowed by the language, it continues to be disallowed. We can already access
+already-initialized members in initialization lists, which means `this` is
+already available, even though it hasn't been completely constructed yet.
+
+
+### What about pass by value methods?
+
+We think they are a logical extension of the mechanism, and would go a long way
+towards making methods as powerful as inline friend functions, with the only
+difference being the call syntax.
+
+One really cool implication of this is that the `this` parameter would be
+move-constructed in the case where the object is an _rvalue_, allowing you to
+treat chained builder methods that return a new object uniformly.
+
+*Example:*
+
+```cpp
+class string_builder {
+  std::string s;
+
+  operator std::string (T this) {
+    return std::move(s);
+  }
+  string_builder operator*(T this, int n) {
+    assert(n > 0);
+    
+    s.reserve(s.size() * n);
+    auto const size = s.size();
+    for (auto i = 0; i < n; ++i) {
+      s.append(s, 0, size);
+    }
+    return this;
+  }
+  string_builder bop(T this) {
+    s.append("bop");
+    return this;
+  }
+};
+
+// this is optimally efficient as far as allocations go
+std::string const x = (string_builder{{"asdf"}} * 5).bop();
+```
+
+Of course, implementing this example with templated `this` methods would have
+been more efficient due to just having fewer objects, but we got rid of all
+references in the program!
+
+
+### `this` as a reference
+
+This paper turns `this` into a reference on an opt-in basis, which is in line
+with the existing guidance that never-null pointers should be references if at
+all possible, and in this case, it is possible.
+
+We believe there would be no confusion, as in all cases, the value category of
+`this` is stated plain as day in the parameter list, which is on the very same
+screen.
+
+Teaching also becomes easier, as the meaning of what a 'const method' is becomes
+more obvious to students.
```

Comment:
Maybe this is a good place to mention the python similarity and/or mental model point in the design decisions section?

Reply:
I think rust is a better bet - python doesn't have the complexity.

Reply:
Kept both in.